core/src/jmh/java/io/grpc/internal/SerializingExecutorBenchmark.java

```diff
@@ -0,0 +1,133 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * SerializingExecutor benchmark.
+ */
+@State(Scope.Benchmark)
+public class SerializingExecutorBenchmark {
+
+  @Param({"CLASSIC", "SINGLE_PRODUCER"})
+  public Type type;
+
+  public enum Type {
+    CLASSIC, SINGLE_PRODUCER
+  }
+
+  private ExecutorService executorService;
+  private Executor executor;
+
+  private static class IncrRunnable implements Runnable {
+    int val;
+
+    @Override
+    public void run() {
+      val++;
+    }
+  }
+
+  private final IncrRunnable incrRunnable = new IncrRunnable();
+
+  private final Phaser phaser = new Phaser(2);
+  private final Runnable phaserRunnable = new Runnable() {
+    @Override
+    public void run() {
+      phaser.arrive();
+    }
+  };
+
+  @Setup
+  public void setUp() {
+    executorService = Executors.newSingleThreadExecutor();
+    switch (type) {
+      case CLASSIC:
+        executor = new SerializingExecutor(executorService);
+        break;
+      case SINGLE_PRODUCER:
+        executor = new SingleProducerSerializingExecutor(executorService);
+        break;
+      default:
+        throw new AssertionError();
+    }
+  }
+
+  @TearDown
+  public void tearDown() throws Exception {
+    executorService.shutdownNow();
+    if (!executorService.awaitTermination(1, TimeUnit.SECONDS)) {
+      throw new RuntimeException("executor failed to shut down in a timely fashion");
+    }
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.SampleTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public void oneRunnableLatency() throws Exception {
+    executor.execute(phaserRunnable);
+    phaser.arriveAndAwaitAdvance();
+  }
+
+  /**
+   * Queue many runnables, to better see queuing/consumption cost instead of just context switch.
+   */
+  @Benchmark
+  @BenchmarkMode(Mode.SampleTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public void manyRunnables() throws Exception {
+    incrRunnable.val = 0;
+    for (int i = 0; i < 500; i++) {
```

Comment:
Meh. I don't think you should give too much credence to the benchmark and if a really long queue forms, the results are even more trash than they already are.

The purpose of this benchmark class is not really to say that the implementation is X% faster, but that the code behaves roughly how we would expect in these cases to give confidence that it would behave how we would expect in the real world. I don't really care if it is faster or slower, as long as it is in the same ballpark. Given that very subtle timing effects that have no impact in real life can produce huge results, consider everything here trash. The benchmark is probably most valuable in that it pushes the implementation, which should weed out many multithreading bugs.